### PR TITLE
[OpInfo] Add multi_head_attention_forward

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100153

### <samp>🤖 Generated by Copilot at 8f8d620</samp>

This pull request improves the testing of the `nn.functional.multi_head_attention_forward` function by adding it to the `OpInfo` framework, adjusting the tolerance and skipping criteria for some test cases, and restricting the dtype for the `MetaProgrammingSystem` tests. These changes aim to address the randomness and numerical precision issues of the function.

### CODE REVIEW

**Comment by qqaatw** [comment]
@zou3519 @ngimel @mruberry Could you take a look at this PR? thank you!

**Comment by qqaatw** [comment]
Hi @zou3519 @ngimel @mruberry, could you review this PR?

**Comment by ngimel** [comment]
cc @drisspg, @jbschlosser

**Comment by malfet** [comment]
@drisspg can you please have a look at it?

**Comment by qqaatw** [comment]
@drisspg, @jbschlosser Hi, can you please review this PR?

**Review by drisspg — APPROVED** [review]
One small comment but looks good. Also think you might need to rebase this pr

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -8039,6 +8039,81 @@ def sample_inputs_max_unpool_grad(op_info, device, dtype, requires_grad, **kwarg
...
+    if requires_grad:
+        # backward tests would take too long to complete, causing the job timeout.
+        bsz = 2
+        is_batcheds = (True,)
+        use_separate_proj_weights = (False,)
+        emb_sizes = (2,)
+        src_lens = (XS,)
+        tgt_lens = (XS,)
+        heads = (2,)
+        dropouts = (0.5,)
```
[the remark this answers is not in the document]
Comment (qqaatw) [reply torch/testing/_internal/common_methods_invocations.py:8054]: The reason of it is that `nn.functional.scaled_dot_product_attention` tests non-zero dropouts as well. What do you think?

[the remark this answers is not in the document]
Comment (drisspg) [reply torch/testing/_internal/common_methods_invocations.py:8054]: ohh iw as just saying I think we should do ` dropouts = (0.0, 0.5,)`

[the remark this answers is not in the document]
Comment (qqaatw) [reply torch/testing/_internal/common_methods_invocations.py:8054]: oh ok, got it :)

**Comment by drisspg** [comment]
I would let this run all the way and if there is no mha failures I'll force merge, seems like CI is having some struggles

**Comment by qqaatw** [comment]
@drisspg The failures look unrelated to mha, so I'm going to force merge this.

@pytorchbot merge -f "Failures look unrelated to mha:  https://hud.pytorch.org/pytorch/pytorch/pull/100153?sha=c8b93beb42f1ec378f6a723b67bca2e26ac4b5f9"

**Pull request closed without merging**